Heat solver interface

dbochkov-flexcompute opened the pull request:
Since the old PR (https://github.com/flexcompute/tidy3d/pull/896) became very outdated, I've decided to open a new one.

Major changes/additions (directory structure is up for discussion, just whatever seemed to make sense for now):

### `components/heat_spec.py`: 
All medium classes can now have field `heat_spec` which can be either `SolidSpec` or `FuildSpec` for specifying thermal properties. To remind, there are three categories of properties:
1. optic (permittivity, conductivity, ...) - defined using standard `Medium`, `Lorentz`, etc classes
2. thermal (heat conductivity, capacity, ...) - added through field `heat_spec`.
3. response of optic properties to temperature (perturbations) - added when using `PerturbationMedium` and `PerturbationPoleResidue` classes.

And this is how different options interact with heat solver:
- Structures using medium instances with `heat_spec=None` or `heat_spec=FluidSpec()` will **not** be included in solution domain for the heat equation, and they will **not** be converted to custom mediums after applying temperature fields.
- Structures using non-perturbation medium instances with `heat_spec=SolidSpec(...)` will be included in solution domain for heat equation, but they will **not** be converted to custom mediums after applying temperature fields.
- Structures using perturbation medium instances (`PerturbationMedium` and `PerturbationPoleResidue`) with `heat_spec=SolidSpec(...)` will be included in solution domain for heat equation, and they will be converted to custom mediums after applying temperature fields. (the fullest material description)

### `components/scene.py`: 
Carved out geometric only information from `Simulation` into class `Scene` for easier transfer of geometric information between different solvers. Basically, it contains `structures`, `medium`, `center`, `size` + plotting functionality. 
- Point to discuss: plotting of custom media is dependent on simulation grid in `Simulation`. I think we should still have eps plotting in `Scene` class (along with thermal conductivity plotting, and perhaps other properties in future), so I adjusted the plotting of custom medium to use custom medium sample points + structure bounds instead of simulation grid. But now the question is whether `Simulation` in future should just directly borrow this plotting method or to have its own based on actual simulation grid?
- As discussed, for now we will not make any changes to `Simulation`, only added property `Simulation.scene` and class method `Simulation.from_scene(scene: Scene, **kwargs)` to transferring `Scene` between optic/heat simulation.
- Contrary, `HeatSimulation` does have field `scene: Scene` 

### `components/bc_placement.py`: 
This file defines classes for placement of boundary conditions on complex interfaces (`BCPlacementType=[StructureBoundary, StructureStructureInterface, MediumMediumInterface, etc]`). I realized this should be separated from heat solver components because it can be reused for other future solvers.

### `components/heat/`: 
The remaining of heat components are placed in this directory to keep everything tidy (yes, you see what I did here lol) and they include:
- `boundary.py` boundary conditions specifications: `TemperatureBC`, `HeatFluxBC`, `ConvectionBC` + `HeatBoundarySpec`
- `grid.py` heat grid specification: currently, `UniformHeatGrid` and `DistanceHeatGrid`. I can that this can go either way: we can introduce more classes to give the user more control or, contrary, unify everything into one spec so that the user don't have to think about what to select at all.
- `sim_data.py` defines `HeatSimulationData` which is currently simply `HeatSimulation` instance + temperature `SpatialDataArray`
- `simulation.py` defines `HeatSimulation`
- `source.py` heat source: currently only `UniformHeatSource`
- `viz.py` some parameters for plotting heat information


### Notebooks with demonstrations
There are two draft notebooks demonstrating:
- basic heat solver usage + comparison to analytical solutions https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/heat-solver-draft/docs/source/notebooks/HeatSolverDraft.ipynb
- usage in the context of thermally tuning a ring resonator https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/heat-solver-draft/docs/source/notebooks/ThermallyTunedRingResonatorDraft.ipynb
These are meant to provide more explanation on how every component is interacting with each other


### additional major things off the top of my head to discuss:
- Currently boundary conditions and heat source placement specifications are done through using structures and medium names. For example, `placement = td.StructureStructureInterface(structures=["box", "sphere"])`. This doesn't seem ideal. We could maybe just pass structures/mediums themselves. But that will increase size of `HeatSimulation` instances. Maybe an alternative is to just record hash of passed structures/mediums? But then it would be harder to point user which structure/medium is missing from a heat simulation.

tylerflex commented on tidy3d/components/heat/simulation.py:
```diff
@@ -0,0 +1,760 @@
+
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    @staticmethod
+    def _construct_heat_boundaries(
```
this function is extremely long.. does it make sense to try to break it up into sub functions? It would be a lot of work and potentially more boilerplate / functions being used once but could make it easier to read and potentially more re-usable. I'll leave it up to you.

tylerflex commented on tidy3d/components/heat/source.py:
```diff
@@ -0,0 +1,48 @@
+class HeatSource(ABC, Tidy3dBaseModel):
+    """Abstract heat source."""
+
+    structures: Tuple[str] = pd.Field(
```
I might be mistaken but I think `Tuple[str]` means that it has to be a tuple of only **one** string, eg `("hello",)`.  I think what you might want is `Tuple[str, ...]` but double check me on this.

tylerflex commented on tidy3d/components/heat_spec.py:
```diff
@@ -0,0 +1,51 @@
+
+    capacity: pd.PositiveFloat = pd.Field(
+        title="Heat capacity",
+        description=f"Volumetric heat capacity {SPECIFIC_HEAT_CAPACITY}.",
```
add "in units of" maybe

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -181,6 +182,13 @@ def _validate_nonlinear_spec(cls, val):
+    heat_spec: Optional[HeatSpecType] = pd.Field(
+        None,
+        title="Heat Specification",
+        description="Specification of the medium heat properties.",
```
maybe we can add more description here. For example something like
```
If a heat_spec is added to a medium, it will be registered if the corresponding ``Structure`` is added to a ``HeatSimulation``.  <insert something about how to run heat sim and convert to optical sim>.  Otherwise, the ``heat_spec`` will not affect the running of an optical ``Simulation``.
```
Basically, just give a curious user looking at the medium API the opportunity to learn more about the heat solver.

tylerflex commented on tidy3d/components/scene.py:
```diff
@@ -0,0 +1,1171 @@
+
+
+class Scene(Box):  # pylint:disable=too-many-public-methods
+    """Contains geometric information to use in simulations.
```
Maybe we can make this more clear, such as
```
Contains generic information about the geometry and medium properties common to all types of simulations.
```

tylerflex commented on tidy3d/components/scene.py:
```diff
@@ -0,0 +1,1171 @@
+
+    """ Post-init validators """
+
+    def _post_init_validators(self) -> None:
```
do we need this here? or should it just be in `Simulation`? I forget how this works or if we need it to be defined inside of a scene itself.

tylerflex commented on tidy3d/web/heat.py:
```diff
@@ -0,0 +1,366 @@
+"""Web API for heat simulation"""
```
We'll clearly need to unify the webapis between all of the solvers because this is almost a direct copy of the `Simulation` webapi right? 

I'm working with MC to provide a simple interface where they provide a template for a `Submittable` task (simulation, mode solver, heat simulation) and we implement this template in our webapi and use a single webapi to manage it. We'll see how that looks but that would hopefully eliminate the need for this file.

dbochkov-flexcompute replied:
let's ignore web api stuff for now until stub is merged

tylerflex reviewed (changes requested):
Looks really good @dbochkov-flexcompute. Mostly minor comments and clarifications here.

dbochkov-flexcompute commented on tests/test_components/test_scene.py:
```diff
@@ -0,0 +1,499 @@
+
+SCENE_FULL = td.Scene(
+    size=(8.0, 8.0, 8.0),
+    structures=[
```
good idea. In fact, I can just use `SCENE = SIM_FULL.scene` haha

dbochkov-flexcompute commented on tidy3d/components/bc_placement.py:
```diff
@@ -0,0 +1,106 @@
+"""Defines heat material specifications"""
```
fixed

dbochkov-flexcompute commented on tidy3d/components/bc_placement.py:
```diff
@@ -0,0 +1,106 @@
+        """Error if the same structure is provided twice"""
+        if val[0] == val[1]:
+            raise SetupError(
+                "The same structure is provided twice in :class:.`StructureStructureInterface`."
```
fixed

dbochkov-flexcompute commented on tidy3d/components/heat/grid.py:
```diff
@@ -0,0 +1,100 @@
+        units=MICROMETER,
+    )
+
+    min_edges_per_circumference: pd.PositiveFloat = pd.Field(
```
it's only relevant for `Sphere` and `Cylinder`, as now explained in description

dbochkov-flexcompute commented on tidy3d/components/heat/sim_data.py:
```diff
@@ -0,0 +1,76 @@
+        description="``HeatSimulation`` object describing the problem setup.",
+    )
+
+    temperature_data: SpatialDataArray = pd.Field(
```
resolved

dbochkov-flexcompute commented on tidy3d/components/heat/simulation.py:
```diff
@@ -0,0 +1,760 @@
+        return ax
+
+    # FIXME: probably needs revision
+    def plot_line(self, line: Shapely, plot_params: PlotParams, ax: Ax) -> Ax:
```
made this work

dbochkov-flexcompute commented:
Major revision taking into account all comment. Most significant additions are PolySlab support, Symmetry support, and proper monitor/simulation data classes. It seemed to accomplish that the most natural way would be to create abstract base classes for main components since their optic and heat specification will share many things, I put them in `base_sim/` for now. So, aiming at the following structure eventually for simulation classes (and similarly for other components):
```
      Scene
        |
  (composition)
        |
AbstractSimulation(ABC)------------------
        |                               |
  (inheritance)                   (inheritance)
        |                               |
Simulation(AbstractSimulation)    HeatSimulation(AbstractSimulation)
```

Demonstration notebooks have been updated too:
- https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/heat-solver-draft/docs/source/notebooks/HeatSolverDraft.ipynb
- https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/heat-solver-draft/docs/source/notebooks/ThermallyTunedRingResonatorDraft.ipynb
These are meant to provide more explanation on how every component is interacting with each other


One concerns: we chose to just copy relevant code from `Simulation` to `Scene` for now, thinking that just removing duplicating code would be easier if everything works out. But I start noticing that `Simulation` functionalities start diverging from whatever I have put in `Scene`. This could happen for other components. So, perhaps, we shouldn't wait too long until we actually reimplement `Simulation` and other optic classes based on introduced in this PR `Scene` and higher level abstractions.

momchil-flex commented:
Yes, that sounds important. Let's aim to have that done for 2.5 too. Maybe even do the refactor as soon as we start a 2.5 branch?

momchil-flex commented:
I see, not so straightforward yeah, but in terms of backwards compatibility I think we'll need to do something like point 2. for now, where simulation is still initialized with center, size, medium, and then a scence is maybe initialized as a private attribute or is stored as a cached property or something.

dbochkov-flexcompute commented:
there are currently 4 commits in this PR:
- "heat solver interface": last review state
- "heat solver web api": [new stuff] implementation of heat web api based on stub refactor
- "moving structures/medium back from ...": [new stuff]
-- removes `size` and `center` from `Scene` and makes them automatically calculated properties.
-- renames `UniformHeatGrid` and `DistanceHeatGrid` to `UniformUnstructureGrid` and `DistanceUnstructureGrid` since they could be reused in other solvers
-- makes `AbstractSimulation` be a derivative of `Box`, thus `domain` field is no longer needed. Simulation domains is set as usual through `size` and `center`.
-- add `structures` and `medium` to `AbstractSimulation`. + property `scene` + class method `from_scene`
- "temporarily limiting matplotlib ...": just a temporary commit to make it easier to deploy and test, can be removed at merge if no longer required

tylerflex commented on tidy3d/components/scene.py:
```diff
@@ -0,0 +1,1274 @@
+        mediums = {structure.medium for structure in val}
+        if len(mediums) > MAX_NUM_MEDIUMS:
+            raise SetupError(
+                f"Tidy3d only supports {MAX_NUM_MEDIUMS} distinct mediums."
```
just noticed, this should be Tidy3D (capital D)

tylerflex commented on tidy3d/components/scene.py:
```diff
@@ -0,0 +1,1274 @@
+
+    @cached_property
+    def bounds(self) -> Bound:
+        """Automatically defined scene's bounds based on present structures. Infinite dimensions
```
seems like a reasonable way to do this

tylerflex commented on tidy3d/components/scene.py:
```diff
@@ -0,0 +1,1274 @@
+        Returns
+        -------
+        Dict[:class:`.AbstractMedium`, int]
+            Mapping between distinct mediums to index in simulation.
```
simulation -> scene (everywhere in file)

tylerflex commented on tidy3d/components/scene.py:
```diff
@@ -0,0 +1,1274 @@
+
+    """ Plotting General """
+
+    @staticmethod
```
I think it's ok for now, but eventually we could probably put a lot of these static methods in viz.py or some `components/scene/viz.py` if we wanted to separate them out.

tylerflex commented on tidy3d/components/simulation.py:
```diff
@@ -3217,3 +3218,49 @@ def perturbed_mediums_copy(
         return Simulation.parse_obj(sim_dict)
+
+    @cached_property
+    def scene(self) -> Scene:
```
is this implemented in the super class?

tylerflex replied:
same for `from_scene`? also add from_scene return type

dbochkov-flexcompute replied:
actually, currently `Simulation` doesn't inherit from the new `AbstractSimulation` per our discussion to leave it on its own for a bit (similarly, `SimulationData`, monitors, etc). But I would like to do this transition in the next PR.

tylerflex commented on tidy3d/components/validators.py:
```diff
@@ -182,6 +182,39 @@ def objects_in_sim_bounds(cls, val, values):
+        scene = values.get("scene")
+        if scene is None:
+            raise ValidationError(
+                f"Cannot validate {field_name} because validation of 'HeatSimulation.scene' "
```
quotes around field name

tylerflex commented on tidy3d/web/api/tidy3d_stub.py:
```diff
@@ -19,29 +20,31 @@
 class Tidy3dStub(BaseModel, TaskStub):
 
-    simulation: SimulationType
+    simulation: SimulationType = pd.Field(discriminator="type")
```
eventually we should add some more description here, if only to help other developers understand

tylerflex commented on tidy3d/web/api/webapi.py:
```diff
@@ -273,125 +273,159 @@ def monitor(task_id: TaskId, verbose: bool = True) -> None:
-        else:
-            est_flex_unit = task_info.estFlexUnit
-            if est_flex_unit is not None and est_flex_unit > 0:
+        # Wait for task to finish
```
I won't go through all of this logic, but we should test this pretty thoroughly because the webapi logic tends to break in weird ways sometimes.

dbochkov-flexcompute replied:
good idea, everything seemed to be working, except for download_log. but it was an issue on backend and now is fixed.

tylerflex replied:
have you tried job and batch of heat simulation? or even mixes of simulation and heat sim? it could be interesting to see how that works too. 

In my experience, sometimes the bugs were happening in `web.monitor()` in weird edge cases, for example 
* a task errors (if that happens we generally try the task again and see if it errors twice)
* the timing is somewhat different for whatever reason (task runs before the % done is ready)

There were some improvements earlier this year to fix some of these issues but if this PR changes any of that logic, they could appear again.  Will be good to run a bunch just to ensure they don't pop up again because they can be quite tricky to reproduce.

tylerflex reviewed (changes requested):
Thanks @dbochkov-flexcompute. I think this looks really good. Mostly this is just minor stuff.

dbochkov-flexcompute commented on tidy3d/components/heat/simulation.py:
```diff
@@ -0,0 +1,815 @@
+    grid_spec: HeatGridType = pd.Field(
+        title="Grid Specification",
+        description="Grid specification for heat simulation.",
+        discriminator=TYPE_TAG_STR,
```
My understanding was that we should use discriminator whenever variable type is Union with more than 1 element, so that pydantic doesn't try to initialize provided argument with every element of that Union. Do you know if something changed during pydantic 2 transition?

tylerflex replied:
I was under the impression that something changed with the discriminator recently, but I think I might be mistaken. let's leave it in.

dbochkov-flexcompute commented on tidy3d/components/heat/simulation.py:
```diff
@@ -0,0 +1,815 @@
+        return val
+
+    @pd.validator("size", always=True)
+    def check_zero_dim_domain(cls, val, values):
```
support is planned after initial release

dbochkov-flexcompute commented on tidy3d/components/heat/simulation.py:
```diff
@@ -0,0 +1,815 @@
+        return ax
+
+    @classmethod
+    def from_scene(cls, scene: Scene, **kwargs):
```
redefining it here to add example in docstring

dbochkov-flexcompute commented on tidy3d/components/heat/simulation.py:
```diff
@@ -0,0 +1,815 @@
+
+        return med_to_bc_spec
+
+    @staticmethod
```
it will eventually go to `components/unstructured_sim`

tylerflex replied:
a lot of plotting logic probably can be separated into other files, since it seems like a lot of them are just `staticmethod`s. But we can do this at a later time, I think.

dbochkov-flexcompute commented on tidy3d/components/bc_placement.py:
it will eventually go to `components/unstructured_sim`

dbochkov-flexcompute commented on tidy3d/components/heat/data/sim_data.py:
```diff
@@ -0,0 +1,232 @@
+
+    @equal_aspect
+    @add_ax_if_none
+    def plot_data(
```
yeah, I was hoping to just have a common implementation, but there were a bit too many specific things for each simulation type. Probably just leave like that for now and revisit after initial release

dbochkov-flexcompute commented:
addressed comments + added a beta stage disclaimer in heat solver webapi:
> Tidy3D's heat solver is currently in the beta stage. All heat simulations are charged a flat fee of 0.025 FlexCredit.

momchil-flex reviewed (commented):
I'm pretty lost here so @dbochkov can you at least help me understand the current state and the plan for the general unification of things?

For example: HeatSimulation inherits from AbstractSimulation, but Simulation does not. Similarly HeatSimulationData inherits from AbstractSimulationData, but SimulationData does not. And while AbstractMonitorData has been introduced, both MonitorData and HeatMonitorData just inherit from Dataset.

Is there currently any duplicated code between Scene and Simulation?

I propose clarifying how everything is meant to work and putting the next few days of effort into actually merging this PR into 2.5 in a way that will avoid duplicated code and further code divergence. Then we can review/fix smaller details.

dbochkov-flexcompute commented:
> I'm pretty lost here so @dbochkov can you at least help me understand the current state and the plan for the general unification of things?
> 
> For example: HeatSimulation inherits from AbstractSimulation, but Simulation does not. Similarly HeatSimulationData inherits from AbstractSimulationData, but SimulationData does not. And while AbstractMonitorData has been introduced, both MonitorData and HeatMonitorData just inherit from Dataset.
> 
> Is there currently any duplicated code between Scene and Simulation?
> 
> I propose clarifying how everything is meant to work and putting the next few days of effort into actually merging this PR into 2.5 in a way that will avoid duplicated code and further code divergence. Then we can review/fix smaller details.

At this stage `HeatMonitorData` should inherit from `AbstractMonitorData`, but `MonitorData` not. I'll fix that. 

So, the idea was to only have heat classes inheriting new `Abstract*` classes to test the new code organization while keeping optic classes untounched, in case something doesn't work out and we need to rethink it. I planned to do the removal of the duplicated code and switching optic classes to `Abstract*` classes in a separate PR. Are you suggesting just to go ahead and do that in this PR? I guess that would have the benefit of double-checking that nothing weird happens when proceeding with this unification.

dbochkov-flexcompute commented:
Otherwise, we seemed to have converged on pretty much everything in this PR. Just need to test out `Batch` with failing heat tasks

momchil-flex commented:
I think separate PR is fine. Just sooner rather than later. Generally as far as I'm concerned we could merge this soon (after you rebase again and answer all current comments, etc.) And just get it out there so that we can start getting some real user feedback and start fixing things. :)

momchil-flex commented:
Is it currently possible for a user to have a look at the mesh that was used after a solver run?

dbochkov-flexcompute commented:
> Is it currently possible for a user to have a look at the mesh that was used after a solver run? I guess they should at least be able to download the mesh file even if we don't provide any gmesh-wrapped functionality because of their GPL license.

yes, it's not ideal currently, but you can directly download mesh ("mesh.cgns") or solution+mesh ("volume_proc0.vtu") using `web.core.s3utils.download_file` to visualize in thirdparty tools like paraview or pyvista. 

This user experience can definitely be improved. Probably, once we have heat monitors that return data on the original unstructured grid instead of interpolated to an equivalent Cartesian one. One idea could be to introduce something like `MeshMonitor` that only return mesh in the selected region. And if a simulation contains only such monitors, then only meshing is performed and solver step is skipped. This will provide a way to inspect and iterate on mesh before actually solving heat equation. Or just provide meshing-only "solver" similar to mode solver.

dbochkov-flexcompute commented on requirements/basic.txt:
```diff
@@ -6,7 +6,7 @@ importlib-metadata>=6.0.0
 h5py>=3.0.0
 rich<12.6.0 # note: rich >= 12.6 adds double progressbars
-matplotlib
+matplotlib<=3.7.2
```
removed
